Create a PR to the upstream repo

jeong-jaeho opened the pull request:
### Duke (Linus chatbot)
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
* ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. download it from [here](https://github.com/nus-cs2103-AY2324S2/ip.git).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [ ] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void `main`(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

jeong-jaeho committed:
Added Level-0: Rename, Greet, Exit

jeong-jaeho committed:
Add Scanner for user input and exiting with "bye".

jeong-jaeho committed:
Add, List

jeong-jaeho committed:
Mark as done

jeong-jaeho committed:
ToDos, Events, Deadlines

jeong-jaeho committed:
Test using the I/O redirection technique

jeong-jaeho committed:
Handle Errors

jeong-jaeho committed:
Add support for deleting tasks from the list.

cheahTJ commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,133 @@
+import java.util.*;
```
Hi Jaeho!! I feel it would be good to import the specific classes you need from the java util class as it is a good practice to avoid using the wildcard (`*`) :)

jeong-jaeho replied:
Thanks for the suggestion!

cheahTJ commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,14 @@
+    }
+
+    @Override
+    public String toString() {
```
I like how you have overriden the `toString()` method to make it specific to the deadline class. Remember also to include in your javadoc for this method to provide a justification as to why you have overriden the `toString()` method 👍

cheahTJ commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,133 @@
+                    // Check if the input string is long enough
+                    if (input.length() <= 5) {
+                        throw new DukeException("Please specify the description of the todo task, " +
+                                "starting from one whitespace away from the keyword 'todo'" +
```
Really good practice of the correct indentation here. It made string message so easy to read.

ondretann replied:
I agree with this. Correct indentation for "if else" has helped with readability for the logic of the code too.

cheahTJ commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,133 @@
+                } else if (input.startsWith("deadline")) {
+                    String[] substrings = input.split(" /by ");
+
+                    if (substrings[0].length() <= 9) {
```
Really good break (empty space) between each logical unit. It makes it so easy to understand your code!

cheahTJ commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,133 @@
+import java.util.*;
+
```
I really like how you have managed to have the appropriate standard spacing (e.g. before a '{') in your code. Continue the good effort!

ondretann commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,31 @@
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
I like how you named your boolean variable! :)

ondretann reviewed (commented):
LGTM

jeong-jaeho committed:
Add Save feature

jeong-jaeho committed:
Merge branch 'branch-Level-7'

jeong-jaeho committed:
Add dates

jeong-jaeho committed:
Merge branch 'branch-Level-8'

jeong-jaeho committed:
Add more OOP

jeong-jaeho committed:
Divide classes into packages

jeong-jaeho committed:
Merge branch 'add-gradle-support'

jeong-jaeho committed:
Add JUnit

jeong-jaeho committed:
Create a JAR File

jeong-jaeho committed:
Add find

jeong-jaeho committed:
Remove potential merge conflict parts

jeong-jaeho committed:
Add JavaDoc comments

jeong-jaeho committed:
Tweak the code to comply with a coding standard

jeong-jaeho committed:
Merge branch 'branch-A-CodingStandard'

jeong-jaeho committed:
Merge branch 'branch-A-JavaDoc'

jeong-jaeho committed:
Resolve merge conflict

jeong-jaeho committed:
Add intermediate commit

jeong-jaeho committed:
Add intermediate commit again

jeong-jaeho committed:
Add a GUI to the chatbot

jeong-jaeho committed:
Merge branch 'branch-Level-10'

jeong-jaeho committed:
Adjust gradle specifications

The entry point of running application is no longer Linus.java, but
it is Launcher.java.

Let's,
* update the mainClass to Launcher.java

jeong-jaeho committed:
Use Assertions

TaskList.java needs Assertions to check if index used in its methods
is within the bounds of the taskList.

Parser.java also needs Assertions to check if the find keyword is
not empty.

These Assertions are needed to ensure application doesn't operate
on wrong inputs.

jeong-jaeho committed:
Improve code quality

The code has magic literals. The code also has deep nesting.
The code also has duplicate conditional fragments.

Let's,
* apply SLAP to resolve deep nesting
* use refactoring to resolve magic literals and duplicates

jeong-jaeho committed:
Merge pull request #2 from jeong-jaeho/branch-A-Assertions

Add Assertions

jeong-jaeho committed:
Merge branch 'master' into branch-A-CodeQuality

jeong-jaeho committed:
Merge branch 'branch-A-CodeQuality'

jeong-jaeho committed:
Add an extension for providing help message

jeong-jaeho committed:
Merge branch 'branch-C-Help'

jeong-jaeho committed:
Add a representative screenshot of the product

jeong-jaeho committed:
Add a User Guide

jeong-jaeho committed:
Edit the main README.md too

jeong-jaeho committed:
Edit implementation

jeong-jaeho committed:
Release the product

jeong-jaeho committed:
Finalize editing
